# Exploring eBay Car Sales Data

In this project, we will be working with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. You can download the dataset from [here](https://data.world/data-society/used-cars-data).

The dataset contains 50,000 datapoints, with the data dictionary as below:
- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the car listings.

In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

As seen from above, our dataset contains a few int64 columns and a few object (string) type columns. Some of these string columns, will be converted to numeric types when we perform data cleaning.

Also we can note that there are a few columns with NaN values. We will be dealing with those columns as well, when we perform data cleaning.

We can also observe, that the column names are in CamelCase instead of Python's preferred snakecase.

Let us first change these column names to snakecase. We will also be rewording some of the names, so that they match the data dictionary description.

### Cleaning Column Names

In [4]:
columns_copy = autos.columns
new_columns = []
for c in columns_copy:
    if c == "yearOfRegistration":
        c = "registration_year"
    elif c == "monthOfRegistration":
        c = "registration_month"
    elif c == "notRepairedDamage":
        c = "unrepaired_damage"
    elif c == "dateCreated":
        c = "ad_created"
    else:
        new_col = ""
        for s in c:
            if ord(s) >= 65 and ord(s) <= 90:
                new_col += "_" + s.lower()
            else:
                new_col += s
        c = new_col
    new_columns.append(c)
    
autos.columns = new_columns
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_p_s', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


Having modified the column names, let us now analyze the first 5 rows of our autos dataset.

In [5]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model   odometer  registration_month fuel_type  \
0    manuell        158  andere  150,000km                   3       lpg   
1  automatik        286     7er  150,000km                   6    benzin   
2    manuell        102    golf   70,000km                   7    benzin   
3  automatik         71  fortwo   70,000km                   6    benzin   
4    manuell          0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Now let us explore the dataset more so as to find out data in columns which might be redundant (i.e a column having the same values thereby not serving any purpose for our project), or numeric data stored as string.

In [6]:
autos.describe(include = "all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-08 10:40:35  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Observing the output of the descibe method, we can see that the columns "seller" and "offertype" have most rows having the same value (seller --> "privat" and offertype --> "Angebot"). So these columns can be removed as redundant data will serve no purpose for data analysis.

Details regarding the "nrofpictures" column is not clear from the above description, let us analyze it further.

In [7]:
autos["nr_of_pictures"].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

This shows the "nrofpictures" column has "0" value for every row, so this row can be removed too.

Now let us focus on columns storing numerical data as strings. Price and Odometer are both stored as strings, this will make working with them harder. Let us convert these to numeric types.

But before doing that we would like to explore their data and identify patterns in them.

### Cleaning Column Values

#### Numeric Columns

In [8]:
autos["price"].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

"price" column has many unique values, all of which contain '$',',' characters.

Let us now remove these non-digit characters and then we will convert these to int64 dtype.

In [9]:
autos["price"] = autos["price"].str.replace(",", "")
autos["price"] = autos["price"].str.replace("$", "")
autos["price"] = autos["price"].astype(int)

autos["price"].dtypes

dtype('int64')

Let us follow the same steps for "odometer" column.

In [10]:
autos["odometer"].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [11]:
autos["odometer"] = autos["odometer"].str.replace(",", "")
autos["odometer"] = autos["odometer"].str.replace("km", "")
autos["odometer"] = autos["odometer"].astype(int)

autos["odometer"].dtypes

dtype('int64')

We would also like to rename the column name from "odometer" to "odometer_km" to help us remember that the cleaned_data contains values in km.

In [12]:
autos = autos.rename({"odometer":"odometer_km"}, axis = 1)
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_p_s', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

Now that we have converted "price" and "odometer" columns to numeric types, let us analyze the values for these two columns to check whether we are able to find outliers (unrealistically high or low values).

In [13]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

"price" column has 0 and 100M as the min and max values. Clearly these data look flawed.
Let us apply `value_counts()` in combination of `sort_index` and `head` to further analyze the min and max values.

In [14]:
autos["price"].value_counts().sort_index(ascending=False).head(10)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64

As can be seen, cars priced more than 10M are unrealistic. But in order for us to decide which rows to remove based on the high/unrealistic prices, we need to further investigate the data.

In [15]:
print(autos.loc[autos["price"] > 1000000, ["name","price","odometer_km"]])

                                                    name     price  \
2897    Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000  11111111   
7814                                         Ferrari_F40   1300000   
11137  suche_maserati_3200_gt_Zustand_unwichtig_laufe...  10000000   
22947                         Bmw_530d_zum_ausschlachten   1234566   
24384                            Schlachte_Golf_3_gt_tdi  11111111   
27371                                         Fiat_Punto  12345678   
39377                        Tausche_volvo_v40_gegen_van  12345678   
39705                        Tausch_gegen_gleichwertiges  99999999   
42221                                  Leasinguebernahme  27322222   
47598  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...  12345678   
47634                                        Ferrari_FXX   3890000   

       odometer_km  
2897         50000  
7814         50000  
11137       100000  
22947       150000  
24384       150000  
27371       150000  
39377       

We notice that 'Ferrari_FXX', a sports car is priced at around 4M USD, which is realistic. For our analysis purposes, we will consider all cars priced more than 5M as unrealistic and remove them from our dataset.
Now let us identify the lower bound.

In [16]:
autos["price"].value_counts().sort_index(ascending=True).head(50)

0      1421
1       156
2         3
3         1
5         2
8         1
9         1
10        7
11        2
12        3
13        2
14        1
15        2
17        3
18        1
20        4
25        5
29        1
30        7
35        1
40        6
45        4
47        1
49        4
50       49
55        2
59        1
60        9
65        5
66        1
70       10
75        5
79        1
80       15
89        1
90        5
99       19
100     134
110       3
111       2
115       2
117       1
120      39
122       1
125       8
129       1
130      15
135       1
139       1
140       9
Name: price, dtype: int64

In [17]:
print(autos.loc[autos["price"] < 100, ["name","price","odometer_km"]])

                                                    name  price  odometer_km
25     Ford_escort_kombi_an_bastler_mit_ghia_ausstattung     90       150000
27                Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE      0       150000
30                            Peugeot_206_Unfallfahrzeug     80       150000
55                        Mercedes_E320_AMG_zu_Tauschen!      1       125000
64                    Autotransport__Abschlepp_Schlepper     40       150000
...                                                  ...    ...          ...
49884               Audi_a6_2.5l__Schnaeppchen_nur_heute      0       150000
49943                                         Opel_astra      0       150000
49960            Ford_KA_zu_verschenken_***Reserviert***      0       150000
49974  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...      0       150000
49984  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...      0       150000

[1762 rows x 3 columns]


We will ignore those cars priced at less than 100USD, even though they are used, as these seem unrealistic.

Removing Unrealistic Prices
***

In [18]:
autos = autos[autos["price"].between(100, 5000000)]
autos["price"].describe()

count    4.823000e+04
mean     6.125042e+03
std      2.286169e+04
min      1.000000e+02
25%      1.250000e+03
50%      3.000000e+03
75%      7.499000e+03
max      3.890000e+06
Name: price, dtype: float64

Let us follow the same approach to identify the outliers for "odometer_km" column.

In [19]:
autos["odometer_km"].describe()

count     48230.000000
mean     125916.545718
std       39546.680447
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [20]:
autos["odometer_km"].value_counts()

150000    31215
125000     5038
100000     2101
90000      1733
80000      1412
70000      1214
60000      1153
50000      1010
40000       814
30000       777
5000        761
20000       757
10000       245
Name: odometer_km, dtype: int64

Data for "odometer_km" column seems realistic, so no need to change anything here.

#### Date Columns

Now, let us shift our focus on date columns "datecrawled", "registration_year", "registration_month", "ad_created" and "lastseen".

Of these pandas has stored "datecrawled", "ad_created" and "lastseen" as string columns. In order to understand the data quantitatively we need to convert these to numeric types.

In [21]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

All the above columns are stored as datetime stamps.
We are only intersted in the date part of the datetime value. So we will extract the date part from the string object and assign it back the respective columns.

In [22]:
autos["date_crawled"] = autos["date_crawled"].str[:10]
autos["ad_created"] = autos["ad_created"].str[:10]
autos["last_seen"] = autos["last_seen"].str[:10]

In [23]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled  ad_created   last_seen
0   2016-03-26  2016-03-26  2016-04-06
1   2016-04-04  2016-04-04  2016-04-06
2   2016-03-26  2016-03-26  2016-04-06
3   2016-03-12  2016-03-12  2016-03-15
4   2016-04-01  2016-04-01  2016-04-01

Now that we have fetched the date parts for the above columns, we will explore the distribution of values in each of these columns.

In [24]:
autos["date_crawled"].value_counts().sort_index()

2016-03-05    1223
2016-03-06     677
2016-03-07    1739
2016-03-08    1600
2016-03-09    1592
2016-03-10    1557
2016-03-11    1572
2016-03-12    1780
2016-03-13     756
2016-03-14    1768
2016-03-15    1655
2016-03-16    1421
2016-03-17    1520
2016-03-18     622
2016-03-19    1675
2016-03-20    1823
2016-03-21    1795
2016-03-22    1587
2016-03-23    1557
2016-03-24    1420
2016-03-25    1519
2016-03-26    1558
2016-03-27    1501
2016-03-28    1686
2016-03-29    1646
2016-03-30    1627
2016-03-31    1536
2016-04-01    1625
2016-04-02    1717
2016-04-03    1862
2016-04-04    1764
2016-04-05     630
2016-04-06     153
2016-04-07      67
Name: date_crawled, dtype: int64

For the date_crawled column the values lie between 5th Mar, 2016 to 7th Apr, 2016. This shows that the web-scraping happened between the above two dates.

Also, as on 5th Mar, 2016 there were only 1,223 car listings on the website (only clean data considered).

In [25]:
autos["ad_created"].value_counts().sort_index().head(50)

2015-06-11       1
2015-08-10       1
2015-09-09       1
2015-11-10       1
2015-12-05       1
2015-12-30       1
2016-01-03       1
2016-01-07       1
2016-01-10       2
2016-01-13       1
2016-01-14       1
2016-01-16       1
2016-01-22       1
2016-01-27       3
2016-01-29       1
2016-02-01       1
2016-02-02       2
2016-02-05       2
2016-02-07       1
2016-02-08       1
2016-02-09       1
2016-02-11       1
2016-02-12       2
2016-02-14       2
2016-02-16       1
2016-02-17       1
2016-02-18       2
2016-02-19       3
2016-02-20       2
2016-02-21       3
2016-02-22       1
2016-02-23       4
2016-02-24       2
2016-02-25       3
2016-02-26       2
2016-02-27       6
2016-02-28      10
2016-02-29       8
2016-03-01       5
2016-03-02       5
2016-03-03      42
2016-03-04      72
2016-03-05    1105
2016-03-06     738
2016-03-07    1678
2016-03-08    1600
2016-03-09    1596
2016-03-10    1543
2016-03-11    1587
2016-03-12    1772
Name: ad_created, dtype: int64

We can see from above that the platform (website) really started to attract traffic from 5th Mar, 2016 onwards.

In [26]:
autos["last_seen"].describe()

count          48230
unique            34
top       2016-04-06
freq           10705
Name: last_seen, dtype: object

The 'last_seen' date tells us the day on which a car was sold, i.e our crawler saw that particular car listing for the last time on that date.

*Note* that last_seen could also be interpreted as the date on which the seller removed their listing from the platform, but for our analysis purposes we will interpret it as date on which the car got sold.

We can see from above that most cars (10,705) got bought on 6th Apr, 2016.

In [27]:
autos["last_seen"].value_counts().sort_index()

2016-03-05       52
2016-03-06      208
2016-03-07      262
2016-03-08      353
2016-03-09      462
2016-03-10      513
2016-03-11      598
2016-03-12     1147
2016-03-13      428
2016-03-14      609
2016-03-15      765
2016-03-16      793
2016-03-17     1355
2016-03-18      353
2016-03-19      760
2016-03-20      996
2016-03-21      991
2016-03-22     1030
2016-03-23      896
2016-03-24      953
2016-03-25      921
2016-03-26      804
2016-03-27      750
2016-03-28     1006
2016-03-29     1076
2016-03-30     1191
2016-03-31     1149
2016-04-01     1102
2016-04-02     1201
2016-04-03     1212
2016-04-04     1183
2016-04-05     6033
2016-04-06    10705
2016-04-07     6373
Name: last_seen, dtype: int64

We stopped crawling after 7th Apr, 2016, as the last date on our date_crawled column shows. So we will not consider the 6,373 cars having '2016-04-07' date as those cars being sold.
Although there might be many cars of these that might have actually been sold on that day.

In [28]:
autos["registration_year"].describe()

count    48230.000000
mean      2004.729795
std         87.892058
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The mean value for the registration_year is 2004, showing that on average a car listed was registered in 2004.
We can also see that min value is 0 and max value is 9999, showing flawed data. We will perform data cleaning exercise on this column as we did for the numeric columns.

In [29]:
autos["registration_year"].value_counts().sort_index(ascending=False).head(15)

9999       3
9000       1
8888       1
6200       1
5911       1
5000       3
4800       1
4500       1
4100       1
2800       1
2019       2
2018     468
2017    1383
2016    1202
2015     380
Name: registration_year, dtype: int64

In [30]:
autos["registration_year"].value_counts().sort_index().head(20)

1000    1
1001    1
1111    1
1800    2
1910    2
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
1950    1
1951    2
1952    1
1953    1
1954    2
Name: registration_year, dtype: int64

Analyzing the some of the most frequent and least frequent 'registration_year' data we see that, years 2017 and beyond are incorrect as a car cannot be registered after it being listed. Determining the earliest valid year is more difficult.

Any car with a registration during the early decades of 1900's will be valid, but for the sake of our project we will ignore all registration year values before 1950 as invalid.

Along with this we will also remove those listings having 'registration_year', 'registration_month' combination that is in excess of (2016, 5).

In [31]:
autos = autos[autos["registration_year"].between(1950, 2016)]

In [32]:
autos["registration_year"].value_counts(normalize=True)

2000    0.066982
2005    0.062817
1999    0.062148
2004    0.058242
2003    0.058113
          ...   
1954    0.000043
1951    0.000043
1953    0.000022
1950    0.000022
1952    0.000022
Name: registration_year, Length: 67, dtype: float64

Now let us try and find out whether there are any listings with (registration_year, registration_month) values higher than (2016, 4).

If we find any, then these will be removed as well.

In [46]:
 autos.loc[(autos["registration_month"] > 4) & (autos["registration_year"] == 2016), ["registration_year","registration_month"]]

registration_year  registration_month
135                 2016                   8
256                 2016                  11
295                 2016                   8
307                 2016                   8
437                 2016                   7
...                  ...                 ...
49547               2016                  10
49852               2016                   7
49876               2016                  10
49919               2016                   8
49938               2016                   6

[612 rows x 2 columns]

In [47]:
 autos = autos.loc[~((autos["registration_month"] > 4) & (autos["registration_year"] == 2016))]

In [48]:
autos.shape

(45729, 20)

### Exploring Price by Brand

Now that our data has been cleaned, we can start to explore and analyze the data.

Below we are intersted in knowing the average price per brand in our car listing data.

In [54]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211026
bmw               0.110783
opel              0.106738
mercedes_benz     0.096963
audi              0.087122
ford              0.069606
renault           0.046994
peugeot           0.029784
fiat              0.025564
seat              0.018107
skoda             0.016510
nissan            0.015264
mazda             0.015198
smart             0.014105
citroen           0.013974
toyota            0.012815
hyundai           0.010037
sonstige_autos    0.009513
volvo             0.009206
mini              0.008835
mitsubishi        0.008157
honda             0.007872
kia               0.007173
alfa_romeo        0.006648
porsche           0.006057
suzuki            0.005948
chevrolet         0.005686
chrysler          0.003499
dacia             0.002646
daihatsu          0.002515
jeep              0.002296
land_rover        0.002121
subaru            0.002121
saab              0.001684
jaguar            0.001553
daewoo            0.001509
trabant           0.001378
r

We will be finding the average price for just the top 6 brands (> 5%).

In [66]:
top6_brands = autos["brand"].value_counts(normalize=True).index[:6]
print(top6_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [69]:
top6_brands_avg_price = {}
for b in top6_brands:
    selected_rows = autos.loc[autos["brand"] == b, "price"]
    top6_brands_avg_price[b] = round(selected_rows.sum() / selected_rows.shape[0])

print(top6_brands_avg_price)

{'volkswagen': 5680.0, 'bmw': 8657.0, 'opel': 3026.0, 'mercedes_benz': 8723.0, 'audi': 9432.0, 'ford': 4122.0}


As can be seen, on average "audi" is the costliest car and "opel" is the cheapest.

We are now curious to find the link between a car's average mileage and it's price.

To make us better understand the connection, we will have to generate average mileage dictionary similar to above.

### Exploring Mileage by Brand

In [73]:
top6_brands_average_mileage = {}
for b in top6_brands:
    selected_rows = autos.loc[autos["brand"] == b, "odometer_km"]
    top6_brands_average_mileage[b] = round(selected_rows.sum() / selected_rows.shape[0])
    
print(top6_brands_average_mileage)

{'volkswagen': 128660.0, 'bmw': 132646.0, 'opel': 129266.0, 'mercedes_benz': 130921.0, 'audi': 129145.0, 'ford': 124172.0}


Now in order to explore the connection between the car prices per brand and their mileages, we will create a new dataframe object with the results of the two dictionary objects obtained from above.

In [75]:
# step 1: create series from dictionary
top6_brands_price_series = pd.Series(top6_brands_avg_price)
# step 2: create dataframe from series
top6_brands_price_df = pd.DataFrame(top6_brands_price_series, columns = ["mean prices"])

top6_brands_price_df["mean mileage"] = pd.Series(top6_brands_average_mileage)

print(top6_brands_price_df)

               mean prices  mean mileage
volkswagen          5680.0      128660.0
bmw                 8657.0      132646.0
opel                3026.0      129266.0
mercedes_benz       8723.0      130921.0
audi                9432.0      129145.0
ford                4122.0      124172.0


We can see, that on average an Audi, BMW, Mercedez get higher prices compared to VolksWagen, Opel or a Ford (even when they have travelled for the same no. of kms)